File: client/src/app/service/task-group.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { environment } from '../environment';
import { Observable } from 'rxjs';

export interface TaskGroup {
    id?: number;
    name: string;
    tasks: any[];
}

@Injectable({ providedIn: 'root' })
export class TaskGroupService {
    private apiUrl = `${environment.apiUrl}task-groups`;

    constructor(private http: HttpClient) {}

    getGroups(): Observable<TaskGroup[]> {
        return this.http.get<TaskGroup[]>(this.apiUrl);
    }

    createGroup(name: string, taskIds: number[]): Observable<TaskGroup> {
        return this.http.post<TaskGroup>(this.apiUrl, { name, taskIds });
    }

    deleteGroup(id: number): Observable<void> {
        return this.http.delete<void>(`${this.apiUrl}/${id}`);
    }

    triggerGroup(id: number, targetDate: Date): Observable<void> {
        return this.http.post<void>(`${this.apiUrl}/${id}/trigger`, { targetDate });
    }

    updateGroup(id: number, name: string, taskIds: number[]): Observable<TaskGroup> {
        return this.http.put<TaskGroup>(`${this.apiUrl}/${id}`, { name, taskIds });
    }
}

File: server/src/main/java/com/couple/taskmanager/service/TaskGroupService.java
package com.couple.taskmanager.service;

import com.couple.taskmanager.model.CTMUser;
import com.couple.taskmanager.model.Task;
import com.couple.taskmanager.model.TaskGroup;
import com.couple.taskmanager.model.dto.TaskGroupDto;
import com.couple.taskmanager.repository.TaskGroupRepository;
import com.couple.taskmanager.repository.TaskRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.Date;
import java.util.HashSet;
import java.util.List;
import java.util.NoSuchElementException;

@Service
public class TaskGroupService {

    @Autowired
    private TaskGroupRepository groupRepository;

    @Autowired
    private TaskRepository taskRepository;

    public List<TaskGroupDto> getGroups(CTMUser user) {
        return groupRepository.findAllByHouseholdId(user.getHousehold().getId())
                .stream().map(TaskGroupDto::new).toList();
    }

    public TaskGroupDto createGroup(String name, List<Long> taskIds, CTMUser user) {
        TaskGroup group = new TaskGroup();
        group.setName(name);
        group.setHousehold(user.getHousehold());

        List<Task> tasks = taskRepository.findAllByIdInAndHouseholdId(new HashSet<>(taskIds), user.getHousehold().getId());
        group.setTasks(tasks);

        return new TaskGroupDto(groupRepository.save(group));
    }

    public void deleteGroup(Long groupId, CTMUser user) {
        TaskGroup group = groupRepository.findByIdAndHouseholdId(groupId, user.getHousehold().getId()).orElseThrow();
        groupRepository.delete(group);
    }

    @Transactional
    public void triggerGroup(Long groupId, Date targetDate, CTMUser user) {
        TaskGroup group = groupRepository.findByIdAndHouseholdId(groupId, user.getHousehold().getId()).orElseThrow();

        // Magically shift all tasks in this library to the target date!
        for (Task task : group.getTasks()) {
            task.setDueDate(targetDate);
            taskRepository.save(task);
        }
    }

    @Transactional
    public TaskGroupDto updateGroup(Long groupId, String name, List<Long> taskIds, CTMUser user) {
        TaskGroup group = groupRepository.findByIdAndHouseholdId(groupId, user.getHousehold().getId())
                .orElseThrow(() -> new NoSuchElementException("Group not found"));

        group.setName(name);
        List<Task> tasks = taskRepository.findAllByIdInAndHouseholdId(new HashSet<>(taskIds), user.getHousehold().getId());
        group.setTasks(tasks);

        return new TaskGroupDto(groupRepository.save(group));
    }
}